File: src/lib/httperror.test.ts
import { describe, expect, it } from 'vitest'
import { errorResponse, statusTitle } from '@/lib/httperror'

describe('statusTitle', () => {
  it('returns RFC 9110 reason phrases (same mapping as Go)', () => {
    expect(statusTitle(400)).toBe('Bad Request')
    expect(statusTitle(401)).toBe('Unauthorized')
    expect(statusTitle(404)).toBe('Not Found')
    expect(statusTitle(409)).toBe('Conflict')
    expect(statusTitle(413)).toBe('Payload Too Large')
    expect(statusTitle(500)).toBe('Internal Server Error')
  })

  it('returns empty string for unknown status', () => {
    expect(statusTitle(999)).toBe('')
  })
})

describe('errorResponse', () => {
  it('builds problem+json response with success/title/status/detail in key order', async () => {
    const res = errorResponse('missing required query parameter: from', 400)
    expect(res.status).toBe(400)
    expect(res.headers.get('content-type')).toContain('application/problem+json')
    const body = await res.json()
    expect(Object.keys(body)).toEqual(['success', 'title', 'status', 'detail'])
    expect(body).toEqual({
      success: false,
      title: 'Bad Request',
      status: 400,
      detail: 'missing required query parameter: from',
    })
  })
})

File: faas/internal/httpx/httperror_test.go
package httpx

import (
	"net/http"
	"testing"
)

// TestStatusTitle 覆盖实际仅用的 6 个已知 status + 未知 status 兜底（docs/20260805-error-response-shape.md「title 映射」）。
func TestStatusTitle(t *testing.T) {
	cases := []struct {
		status int
		want   string
	}{
		{http.StatusBadRequest, "Bad Request"},
		{http.StatusUnauthorized, "Unauthorized"},
		{http.StatusNotFound, "Not Found"},
		{http.StatusConflict, "Conflict"},
		// 413 特例：统一 RFC 9110 新名，覆盖 Go 标准库的 RFC 7231 旧名。
		{http.StatusRequestEntityTooLarge, "Payload Too Large"},
		{http.StatusInternalServerError, "Internal Server Error"},
		{999, ""},
	}
	for _, c := range cases {
		if got := statusTitle(c.status); got != c.want {
			t.Errorf("statusTitle(%d) = %q, want %q", c.status, got, c.want)
		}
	}
}
